index/segment/fs/encoding/fields/data.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Softwarw.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fields
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+)
+
+const (
+	initialDataEncoderLen     = 1024
+	dataEncoderWriteThreshold = 512
+)
+
+// DataWriter writes the data file for stored fields.
+type DataWriter struct {
+	writer io.Writer
+	enc    *encoding.Encoder
+}
+
+// NewDataWriter returns a new DataWriter.
+func NewDataWriter(w io.Writer) *DataWriter {
+	return &DataWriter{
+		writer: w,
+		enc:    encoding.NewEncoder(initialDataEncoderLen),
+	}
+}
+
+func (w *DataWriter) Write(d doc.Document) (int, error) {
+	n := w.enc.PutBytes(d.ID)
+	n += w.enc.PutUvarint(uint64(len(d.Fields)))
+	for _, f := range d.Fields {
+		n += w.enc.PutBytes(f.Name)
+		n += w.enc.PutBytes(f.Value)
+	}
+
+	if w.enc.Len() > dataEncoderWriteThreshold {
```

Comment:
same as the index writer, why not rely on the external callers using buffered writers if they want this behaviour?

Reply:
That's fair, my concern was that if we were given a file handle, this could lead to a lot of syscalls for small writes. Push that decision to the caller sounds good to me though.